src/tsd/PutDataPointRpc.java

```diff
@@ -169,6 +173,20 @@ public void execute(final TSDB tsdb, final HttpQuery query)
       }
     }
     
+    if (synchronous_write) {
+      LOG.warn("Waiting for synchronous write");
+      for (Entry<IncomingDataPoint, Deferred<Object>> result : deferred_results.entrySet()) {
+        try {
+          result.getValue().joinUninterruptibly();
+        } catch (Exception e) {
+          if (show_details) {
+            details.add(this.getHttpDetails(e.getMessage(), result.getKey()));
+          }
+          success--;
+        }
+      }
+    }
+
```

Comment:
We need to avoid the join in here so you should do something like this:
```
    class SyncCB implements Callback<Object, ArrayList<Object>> {
      @Override
      public Object call(final ArrayList<Object> results) throws Exception {
        for (int i = 0; i < results.size(); i++) {
          if (results.get(i) instanceof Exception) {
            if (details != null) {
              final Exception e = (Exception)results.get(i);
              details.add(getHttpDetails(e.getMessage(), dps.get(i)));
            }
            // decrement success counter
          }
        }
        // break out the query.sendReply() code into a method or something that
        // can be called either from this callback or directly if the user doesn't
        // care about sync
        
        return null;
      }
    }
    
    if (sync) {
      Deferred.groupInOrder(deferreds).addCallback(new SyncCB());
    }
```